mutex.go

```diff
@@ -41,102 +40,128 @@ var (
 	storageUnlockURL = defaultStorageUnlockURL
 )
 
-// Lock waits up to duration d for l.Lock() to succeed.
-func Lock(l sync.Locker, d time.Duration) error {
-	done := make(chan struct{}, 1)
-	go func() {
-		l.Lock()
-		done <- struct{}{}
-	}()
-	select {
-	case <-done:
-		return nil
-	case <-time.After(d):
-		return errors.New("lock request timed out")
-	}
-}
-
-// Unlock waits up to duration d for l.Unlock() to succeed.
-func Unlock(l sync.Locker, d time.Duration) error {
-	done := make(chan struct{}, 1)
-	go func() {
-		l.Unlock()
-		done <- struct{}{}
-	}()
-	select {
-	case <-done:
-		return nil
-	case <-time.After(d):
-		return errors.New("unlock request timed out")
-	}
+// ContextLocker provides an extension of the sync.Locker interface.
+type ContextLocker interface {
+	sync.Locker
+	ContextLock(context.Context) error
+	ContextUnlock(context.Context) error
 }
 
 type mutex struct {
-	project string
-	bucket  string
-	object  string
-	client  *http.Client
+	bucket string
+	object string
+	client *http.Client
 }
 
-// Lock waits indefinitely to acquire a global mutex lock.
+var _ ContextLocker = (*mutex)(nil)
+
+// Lock waits indefinitely to acquire a mutex.
 func (m *mutex) Lock() {
+	m.ContextLock(context.Background())
+}
+
+// ContextLock waits indefinitely to acquire a mutex with timeout
+// governed by passed context.
+func (m *mutex) ContextLock(ctx context.Context) error {
 	q := url.Values{
 		"name":              {m.object},
 		"uploadType":        {"media"},
 		"ifGenerationMatch": {"0"},
 	}
 	url := fmt.Sprintf("%s/b/%s/o?%s", storageLockURL, m.bucket, q.Encode())
-	for i := 1; ; i *= 2 {
-		res, err := m.client.Post(url, "plain/text", bytes.NewReader([]byte("1")))
-		if err == nil {
-			res.Body.Close()
-			if res.StatusCode == 200 {
-				return
+	// Starting w/ 10ms, backoff and retry exponentially.
+	for i := 10; ; i *= 2 {
+		done := make(chan bool)
+		go func() {
+			res, err := m.client.Post(url, "plain/text", bytes.NewReader([]byte("1")))
+			if err == nil {
+				res.Body.Close()
+				if res.StatusCode == 200 {
+					done <- true
+					return
+				}
+			}
+			done <- false
+		}()
+		select {
+		case <-time.After(time.Duration(i) * time.Millisecond):
+			continue
```

Comment:
I have to admit it's quite a complicated loop. For instance, what happens with the current `done` channel variable here, on `continue`? and what will the started goroutine do while you're looping over again with the next exponential backoff?

Btw, another issue here is the exponential backoff should be used to compute retry pause only, but it is also used to establish how much time to wait for the GCS reply, which is not the same thing. For instance, you start with `i := 10` which means you'll wait for the `m.client.Post` at most 10 milliseconds, which is very little and unlikely to succeed.

All I'm saying is this loop mixes a few things together and it's now hard to see what happens exactly at any given point. I see why you want to start a new goroutine on each loop iteration but I think it really complicates things a lot. How about replacing that goroutine with the new [Request.WithContext](https://golang.org/pkg/net/http/#Request.WithContext)? Then your loop could look something like:

```go
backoff := 10*time.Millisecond // NOTE: ctx deadline/timeout and backoff are independent 
for {
  req, err := http.NewRequest("POST", url, bytes.NewReader([]byte("1")))
  if err != nil {
    // It is likely a malformed URL; no retry will fix that.
    return err
  }
  req.Header.Set("content-type", "text/plain")
  req = req.WithContext(ctx)
  res, err := m.client.Do(req)
  if err == nil {
    res.Body.Close()
    if res.StatusCode == 200 {
      return nil
    }
  }
  select {
  case <-time.After(backoff):
    backoff *= 2
    continue
  case <-ctx.Done():
    return ctx.Err()
  }
}
```

btw, I believe it's `text/plain`, not `plain/text`.